packages/react-components/react-color-picker-preview/library/src/components/ColorArea/useColorAreaState.ts

```diff
@@ -0,0 +1,120 @@
+import * as React from 'react';
+import { clamp, useEventCallback } from '@fluentui/react-utilities';
+import { useFluent_unstable as useFluent } from '@fluentui/react-shared-contexts';
+import { colorAreaCSSVars } from './useColorAreaStyles.styles';
+import type { ColorAreaState, ColorAreaProps } from './ColorArea.types';
+import { tinycolor, HSVA, Numberify } from '@ctrl/tinycolor';
+import { useColorPickerContextValue_unstable } from '../../contexts/colorPicker';
+
+const { areaXProgressVar, areaYProgressVar, thumbColorVar, mainColorVar } = colorAreaCSSVars;
+
+const MIN = 0;
+const MAX = 100;
+
+export const useColorAreaState_unstable = (state: ColorAreaState, props: ColorAreaProps) => {
+  'use no memo';
+
+  const { targetDocument } = useFluent();
+  const onChangeFromContext = useColorPickerContextValue_unstable(ctx => ctx.requestChange);
+  const { onChange = onChangeFromContext, onMouseDown, onMouseUp } = props;
+
+  const hsvColor = tinycolor(props.color || '#fff').toHsv();
+  const [color, setColor] = React.useState(hsvColor);
+
+  const saturation = Math.round(color.s * 100);
+  const value = Math.round(color.v * 100);
+  const coordinates = { x: saturation, y: value };
+
+  function getCoordinates(event: React.MouseEvent<HTMLDivElement>) {
+    const ref = state.root.ref as React.MutableRefObject<HTMLDivElement>;
+    const rect = ref.current?.getBoundingClientRect();
+    const newX = Math.round(((event.clientX - rect.left) / rect.width) * 100);
+    const newY = 100 - Math.round(((event.clientY - rect.top) / rect.height) * 100);
+
+    return {
+      x: clamp(newX, MIN, MAX),
+      y: clamp(newY, MIN, MAX),
+    };
+  }
+
+  const requestColorChange = useEventCallback((event: React.MouseEvent<HTMLDivElement>) => {
+    const _coordinates = getCoordinates(event);
+    const newColor = { h: hsvColor.h, s: _coordinates.x / 100, v: _coordinates.y / 100, a: 1 };
+    setColor(newColor);
+    onChange?.(event, {
+      type: 'mousemove',
+      event,
+      color: parseColor(newColor),
+    });
+  });
+
+  const _onMouseUp = useEventCallback((event: React.MouseEvent<HTMLDivElement>) => {
+    onMouseUp?.(event);
+    targetDocument?.removeEventListener('mousemove', requestColorChange as unknown as EventListener);
+    targetDocument?.removeEventListener('mouseup', _onMouseUp as unknown as EventListener);
+  });
+
+  const _onMouseDown = useEventCallback((event: React.MouseEvent<HTMLDivElement>) => {
+    event.stopPropagation();
+    event.preventDefault();
+    onMouseDown?.(event);
+    requestColorChange(event);
+    targetDocument?.addEventListener('mousemove', requestColorChange as unknown as EventListener);
+    targetDocument?.addEventListener('mouseup', _onMouseUp as unknown as EventListener);
+  });
```

Comment:
Currently DOM events and React events are mixed which is not great, code below should be a good starting point for cleanup:

```suggestion
  const handleDocumentMouseMove = React.useCallback(
    (event: MouseEvent) => {
      requestColorChange(event);
    },
    [requestColorChange],
  );
  const handleDocumentMouseUp = useEventCallback(() => {
    targetDocument?.removeEventListener('mousemove', handleDocumentMouseMove);
    targetDocument?.removeEventListener('mouseup', handleDocumentMouseUp);
  });

  const _onMouseDown = useEventCallback((event: React.MouseEvent<HTMLDivElement>) => {
    event.stopPropagation();
    event.preventDefault();

    onMouseDown?.(event);
    requestColorChange(event);

    targetDocument?.addEventListener('mousemove', handleDocumentMouseMove);
    targetDocument?.addEventListener('mouseup', handleDocumentMouseUp);
  });
```